### PR TITLE
feat: add QwenImageTrainer for fine-tuning with Adafactor optimizer

### PR DESCRIPTION
**Pull request by kohya-ss**
- [x] basic fine-tuning
- [x] fine-tuning for pruned model
- [x] ~~pre-finetuning for pruned model with LoRA (if possible)~~ Will be addressed in a separate PR in the future

This PR introduces an experimental implementation for the full finetuning of Qwen-Image. Ref #490.

## Options

Please use `qwen_image_train.py`. The basic options are the same as those for Qwen-Image LoRA training, but with the following differences:

### Unsupported Options:

-   FP8 training is not yet supported. `--fp8_base` and `--fp8_scaled` cannot be used (they will be forcibly disabled).
-   `--scale_weight_norms` is not effective.
-   LoRA-related options like `--network_*` will be ignored.

### Added Options:

-   `--full_bf16`: Enables full bfloat16 training, including model weights. This option requires `--mixed_precision bf16`. This reduces VRAM usage by about 20GB but may impact model accuracy as the weights are kept in bfloat16. It is recommended to use this with an optimizer that supports stochastic rounding.  In this repository, Adafactor optimizer with `--fused_backward_pass` option supports stochastic rounding.
-   `--fused_backward_pass`: Reduces VRAM usage during gradient computation by performing the backward pass for each parameter. This is only effective when `--optimizer_type` is set to `adafactor`. Note that Gradient Accumulation is disabled when this option is used.
-   `--mem_eff_save`: Uses a memory-efficient saving method to reduce main memory (RAM) usage during checkpoint saving by about 40GB (for full bf16; down from ~80GB without it). Please note that when saving the optimizer state with `--save_state`, the conventional saving method is used, which will still require about 40GB of main memory.

## Recommended Settings

We are still exploring the optimal settings. The following configurations are just examples, so please adjust them as needed. We welcome your feedback.

If you have ample VRAM, you can use any optimizer of your choice.

For limited VRAM environments (e.g., 48GB or less), you may need to use `--full_bf16`, the Adafactor optimizer, and `--fused_backward_pass`. Here are the recommended options for that case. Please adjust `--lr_warmup_steps` to a value between approximately 10 and 100.

```
--mixed_precision bf16 --full_bf16 --optimizer_type adafactor \
--optimizer_args "relative_step=False" "scale_parameter=False" "warmup_init=False" \
--max_grad_norm 0 --lr_scheduler constant_with_warmup --lr_warmup_steps 10
```

If your VRAM is even more constrained, you can enable block swapping by specifying a value for `--blocks_to_swap`.

Experience with other models suggests that the learning rate may need to be reduced significantly; something in the range of 1e-6 to 1e-5 might be a good place to start.

### CODE REVIEW

**Comment by FurkanGozukara** [comment]
man i need this ty so much

**Comment by kohya-ss** [comment]
Support for pruned model (with fewer than 60 blocks) has been added.

The finetuning script counts the number of transformer blocks in the weights and loads them automatically. It can load even sparse indices. In saving, the indices will be consecutive numbers.

For simplicity, the LoRA training and inference scripts do not automatically detect this. When loading a pruned model, specify the number of blocks, e.g., `--num_layers 30`. These scripts only support dense indices.

**Comment by sdbds** [comment]
So can we now just upload the cache part for fine-tuning? Considering large datasets and cloud storage space, I think we can prepare the cache in advance and be ready.

**Comment by kohya-ss** [comment]
> So can we now just upload the cache part for fine-tuning? Considering large datasets and cloud storage space, I think we can prepare the cache in advance and be ready.

I haven't tested it yet, but it should work with just the .toml file and cache directory.

**Comment by elie-h** [comment]
.toml file and cache directory worked fine

However I couldn't get this to fit on an 80GB card (A100) 
Tried with --full_bf16 and all 59 layers of block swapping

**Comment by FurkanGozukara** [comment]
> .toml file and cache directory worked fine
> 
> However I couldn't get this to fit on an 80GB card (A100) Tried with --full_bf16 and all 59 layers of block swapping

wow this sounds wrong to me.

**Comment by kohya-ss** [comment]
> However I couldn't get this to fit on an 80GB card (A100)
> Tried with --full_bf16 and all 59 layers of block swapping

What optimizer did you use? Could you try AdamW8bit or Adafactor (with/without fused backward pass) optimizer?

**Comment by jferments** [comment]
Does this support multi-GPU setups? I have a 2xRTX 4090 machine (48GB VRAM total), so how would I configure training to use both GPUs? Thanks!

**Comment by kohya-ss** [comment]
> Does this support multi-GPU setups? I have a 2xRTX 4090 machine (48GB VRAM total), so how would I configure training to use both GPUs? Thanks!

Musubi Tuner supports multi-GPU training with Accelerate library. However, I don't have much experience with multi-GPU training, so I don't know if Qwen-Image finetuning will work.

**Comment by sdbds** [comment]
If we add fine-tuning, we mainly need two capabilities:
1. Save state and read it back again, because fine-tuning with large datasets always encounters various situations, such as network issues, dataset issues, and so on.
Although this project also has save state functionality, it seems somewhat incomplete and has some issues.

2. Multi-GPU communication, mainly FSDP. Of course, I think supporting 8 GPUs is sufficient, and multi-node setups might require torchrun, etc.

I think we can refer to the training settings of OmniGen2. Their `save_state` is quite comprehensive, and they directly support EMA.

[load_state](https://github.com/VectorSpaceLab/OmniGen2/blob/0dc40c80d7b357c329fb73422dc7856ed625352c/train.py#L455-L480)

[save_state](https://github.com/VectorSpaceLab/OmniGen2/blob/0dc40c80d7b357c329fb73422dc7856ed625352c/train.py#L638)

[ema](https://github.com/VectorSpaceLab/OmniGen2/blob/0dc40c80d7b357c329fb73422dc7856ed625352c/omnigen2/training_utils.py#L312-L645)

**Comment by kohya-ss** [comment]
> 1. Save state and read it back again, because fine-tuning with large datasets always encounters various situations, such as network issues, dataset issues, and so on.
>    Although this project also has save state functionality, it seems somewhat incomplete and has some issues.

I've tested saving and restoring states with single GPU training and it seems to work well (except for restoring step and epoch numbers). Further testing with multi-GPU training is needed.

Storing EMA requires additional memory, so I think, in practice, it may be better to use Post-Hoc EMA.

> 2\. Multi-GPU communication, mainly FSDP. Of course, I think supporting 8 GPUs is sufficient, and multi-node setups might require torchrun, etc.

I agree. In theory, FSDP should work with accelerate config, but this also needs to be verified.

**Comment by jferments** [comment]
> In theory, FSDP should work with accelerate config, but this also needs to be verified.

I tried using FSDP (both v1 and v2), but was still [getting OOM errors](https://github.com/kohya-ss/musubi-tuner/discussions/498) on a 2xRTX 4090 machine. 

>Musubi Tuner supports multi-GPU training with Accelerate library.

In addition to trying FSDP, I tried using DDP (--num_processes 2 without FSDP installed) with "--blocks_to_swap" set to 24, 36, 48, and 59 blocks (along with all of the other low VRAM options suggested above), and couldn't get any of those to work either.

**Comment by FurkanGozukara** [comment]
> > In theory, FSDP should work with accelerate config, but this also needs to be verified.
> 
> I tried using FSDP (both v1 and v2), but was still [getting OOM errors](https://github.com/kohya-ss/musubi-tuner/discussions/498) on a 2xRTX 4090 machine.
> 
> > Musubi Tuner supports multi-GPU training with Accelerate library.
> 
> In addition to trying FSDP, I tried using DDP (--num_processes 2 without FSDP installed) with "--blocks_to_swap" set to 24, 36, 48, and 59 blocks (along with all of the other low VRAM options suggested above), and couldn't get any of those to work either.

From my FLUX fine tuning experience on multi GPU I can say that, I can do flux fine tuning with as low as 6 GB VRAM with block swap but it requires 80+ GB for fine tuning on even 2x GPU

it never worked below

**Comment by radiceradice** [comment]
I was able to successfully fine-tune QwenImage on an RTX 5090 (using 1000 training images).

Training Conditions:

Image size: Started with higher resolutions and gradually reduced, finally down to 512x512. However, step time did not improve (approx. 30 minutes/step).

Batch size: Tested 1–4, but no significant difference. Eventually used batch=1.

Options tested: --sdpa, --flash_attn, --fused_backward_pass, --gradient_checkpointing
→ None had a major effect.

GPU memory: At batch=1, 512x512, VRAM was nearly 100% (≈32 GB / 32.6 GB) usage throughout.

After enabling --blocks_to_swap:
VRAM usage decreased (overflowing into shared memory), but step time dramatically improved, making training speed practical.

Training Command:
accelerate launch --num_cpu_threads_per_process 1 --mixed_precision bf16 src/musubi_tuner/qwen_image_train.py \
--dit C:\musubi-tuner\models\Qwen-Image\qwen_image_bf16.safetensors \
--vae C:\musubi-tuner\models\Qwen-Image\vae\diffusion_pytorch_model.safetensors \
--text_encoder C:\musubi-tuner\models\Qwen-Image\qwen_2.5_vl_7b.safetensors \
--dataset_config C:\musubi-tuner_data\Qwen-fineturning\dataset.toml \
--sdpa --full_bf16 --gradient_checkpointing \
--timestep_sampling shift --weighting_scheme none --discrete_flow_shift 2.2 \
--optimizer_type adafactor --learning_rate 1e-5 \
--max_data_loader_n_workers 2 --persistent_data_loader_workers \
--max_train_steps 20000 --save_every_n_steps 2000 \
--seed 42 --optimizer_args "relative_step=False" "scale_parameter=False" "warmup_init=False" \
--max_grad_norm 0 --lr_scheduler constant_with_warmup --lr_warmup_steps 10 \
--output_dir C:\musubi-tuner_data\qwen_image_V1 --output_name qwen_fineturning_V1 \
--logging_dir D:\musubi-tuner_data\qwen_image_V1 --log_with tensorboard \
--sample_every_n_steps 2000 --sample_prompts C:\musubi-tuner_data\Qwen-fineturning\Qwen_fineturning\sample.txt \
--save_state --fused_backward_pass --blocks_to_swap 20

Example GPU Status (after enabling --blocks_to_swap):
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 576.80   Driver Version: 576.80   CUDA Version: 12.9                         |
|-----------------------------------------+------------------------+----------------------|
| GPU Name                  Driver-Model | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap | Memory-Usage | GPU-Util  Compute M.          |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 5090      WDDM  |   00000000:02:00.0  On |                  N/A |
| 45%   66C    P1            289W / 575W |   30200MiB / 32607MiB |     63%      Default |
+-----------------------------------------------------------------------------------------+


GPU: RTX 5090 (32 GB)
VRAM usage: 30.2 GB / 32.6 GB
Power consumption: 289 W / 575 W
GPU utilization: 63%

Training Progress (Loss Values):
Initial run (stopped at 20,000 steps):
Step | Loss
2002 | 0.096990
3985 | 0.095866
5991 | 0.095938
7992 | 0.094409
9973 | 0.092730
11999 | 0.092268
13999 | 0.090976
15994 | 0.089573
17991 | 0.087376
20000 | 0.084900

After resuming from 20,000 steps
Step | Loss
20019   | 0.071315
22002   | 0.082977
24001   | 0.080674
25991   | 0.077095
27993   | 0.075320
29997   | 0.074145
30816   | 0.073862
32816   | 0.073862

Training continued up to 35,000 steps.

-No degradation was observed in sample images.
-However, from around 32,000 steps, generated outputs started to show reduced diversity: similar compositions and patterns were repeatedly produced.
-Based on this, I decided to stop at 35,000 steps.

Afterwards, I confirmed that images could be generated in ComfyUI, so I considered this a successful fine-tuning.

However, I feel the generated images still lack some subtle details, so I plan to retry with 1024x1024 training images.

**Comment by FurkanGozukara** [comment]
@kohya-ss i am working on training and is this --fused_backward_pass supported in LoRA in main repo?

if i merge this pull request will it work for lora training?

currently adafactor uses 14.5 GB VRAM and Adamw8bit uses 16.5 GB VRAM minimum for 1328x1328 rank 128

moreover do we currently have stochastic rounding for adafactor? for lora training with adafactor?

thank you so much

these are super important since i will do a full training parameters research

especially stochastic rounding

**Comment by kohya-ss** [comment]
> @kohya-ss i am working on training and is this --fused_backward_pass supported in LoRA in main repo?
> 
> if i merge this pull request will it work for lora training?
> 
> currently adafactor uses 14.5 GB VRAM and Adamw8bit uses 16.5 GB VRAM minimum for 1328x1328 rank 128
> 
> moreover do we currently have stochastic rounding for adafactor? for lora training with adafactor?

Unfortunately, the fused backward pass is only effective with full finetuning.
LoRA has fewer parameters than full finetuning, so the effect of the fused backward pass is limited.
Enabling the fused backward pass probably does not reduce VRAM significantly, so there are no plans to support it at this time.

Stochastic rounding is only effective for full bfloat16 training, but full bfloat16 is generally less accurate, so it is not recommended for LoRA training which has a small number of parameters.

**Comment by sdbds** [comment]
> > @kohya-ss i am working on training and is this --fused_backward_pass supported in LoRA in main repo?
> > if i merge this pull request will it work for lora training?
> > currently adafactor uses 14.5 GB VRAM and Adamw8bit uses 16.5 GB VRAM minimum for 1328x1328 rank 128
> > moreover do we currently have stochastic rounding for adafactor? for lora training with adafactor?
> 
> Unfortunately, the fused backward pass is only effective with full finetuning. LoRA has fewer parameters than full finetuning, so the effect of the fused backward pass is limited. Enabling the fused backward pass probably does not reduce VRAM significantly, so there are no plans to support it at this time.
> 
> Stochastic rounding is only effective for full bfloat16 training, but full bfloat16 is generally less accurate, so it is not recommended for LoRA training which has a small number of parameters.

Actually, previous test results on models like SDXL indicated that full bf16 is more suitable for application in LoRA, as the loss when using full bf16 in LoRA seems much smaller than in fine-tuning...
Using full bf16 in fine-tuning results in significant degradation in image quality (must use mix), but there is none in LoRA...

**Comment by FurkanGozukara** [comment]
> > @kohya-ss i am working on training and is this --fused_backward_pass supported in LoRA in main repo?
> > if i merge this pull request will it work for lora training?
> > currently adafactor uses 14.5 GB VRAM and Adamw8bit uses 16.5 GB VRAM minimum for 1328x1328 rank 128
> > moreover do we currently have stochastic rounding for adafactor? for lora training with adafactor?
> 
> Unfortunately, the fused backward pass is only effective with full finetuning. LoRA has fewer parameters than full finetuning, so the effect of the fused backward pass is limited. Enabling the fused backward pass probably does not reduce VRAM significantly, so there are no plans to support it at this time.
> 
> Stochastic rounding is only effective for full bfloat16 training, but full bfloat16 is generally less accurate, so it is not recommended for LoRA training which has a small number of parameters.

thank you so much so we are at maximum quality

i did a random training yesterday and it works perfect now i will fine tune the parameters

by the way currently the minimum vram is 14.6 GB with full block swap + FP8 for 1328x1328 and lora rank 128

i was hoping lower vram :(

**Comment by elie-h** [comment]
@kohya-ss @FurkanGozukara  sorry I missed the original comment, but here's my config:

```
accelerate launch \
  --num_cpu_threads_per_process 12 \
  --mixed_precision bf16 \
  src/musubi_tuner/qwen_image_train.py \
  --dit models/qwen_image_bf16.safetensors \
  --vae models/qwen_vae.safetensors \
  --text_encoder models/qwen_2.5_vl_7b.safetensors \
  --dataset_config runs/test/dataset.toml \
  --sdpa \
  --gradient_checkpointing \
  --timestep_sampling qwen_shift \
  --weighting_scheme none \
  --discrete_flow_shift 3.0 \
  --optimizer_type adafactor \
  --learning_rate 1e-5 \
  --max_data_loader_n_workers 2 \
  --persistent_data_loader_workers \
  --max_train_steps 20000 \
  --save_every_n_steps 2000 \
  --seed 42 \
  --optimizer_args "relative_step=False" "scale_parameter=False" "warmup_init=False" \
  --max_grad_norm 0 \
  --lr_scheduler constant_with_warmup \
  --lr_warmup_steps 10 \
  --output_dir runs/test/output \
  --output_name test \
  --logging_dir runs/test/output \
  --fused_backward_pass \
  --log_with tensorboard \
  --save_state \
  --blocks_to_swap 59 \
  --sample_every_n_steps 1000 \
  --sample_prompts runs/test/sample.txt
```

I avoided --full_bf16 as I didn't want a drop in quality, that's the only setting that makes it work
I seem to be hitting an OOM with the above even though I have 80GB of VRAM & 128 GB of RAM. I tried adamw & adafactor with and without fused_backward_pass

Any ideas?

**Comment by kohya-ss** [comment]
> I avoided --full_bf16 as I didn't want a drop in quality, that's the only setting that makes it work
> I seem to be hitting an OOM with the above even though I have 80GB of VRAM & 128 GB of RAM. I tried adamw & adafactor with and without fused_backward_pass

The model is 20B, so if you train it with float32, the weights alone will use 80GB of memory. 

Add `--blocks_to_swap` to offload some blocks to the CPU.

**Comment by Pashahlis** [comment]
@kohya-ss This is great, but there isn't yet a utility to extract a Qwen LoRa from a fully finetuned Qwen model is there?

I would love to fully finetune, then extract a LoRa, to see if thats perhaps higher quality than straight up LoRa training.

**Comment by Pashahlis** [comment]
> @kohya-ss This is great, but there isn't yet a utility to extract a Qwen LoRa from a fully finetuned Qwen model is there?
> 
> I would love to fully finetune, then extract a LoRa, to see if thats perhaps higher quality than straight up LoRa training.

@kohya-ss I vibe-coded a qwen-image lora extraction python script using ChatGPT and it seems to work flawlessly. Though no guarantee yet, as my first test seems to have been extremely undertrained, and the second test severely overtrained. Still trying out different LRs. Full-finetuning definitely does seem to need substantially less LR though.

Here is the script: https://www.dropbox.com/scl/fi/znrci4i06h97d5pjbc88h/qwen_image_lora_extraction.py?rlkey=gkwaiz06zhhaqxwsw5dhxwo6d&st=vh4uxuhf&dl=1

**Comment by Pashahlis** [comment]
Nevermind. There is still an issue with it. It did not correctly extract the weights as I found out now after some hours of testing. I'll see if I can fix it before an official lora extraction script is released.

**Comment by Pashahlis** [comment]
Alright, this seems to work now: https://www.dropbox.com/scl/fi/3xrdswqdi40rjpo9bd36q/qwen_image_lora_extraction.py?rlkey=m75g6aijdjkzb9xjf86l7lzhx&st=y8eus9vi&dl=1 Was a key naming issue.

But I find that you need approximately 128 rank for it to have a noticeable effect aka a 2gb big LoRa file. But that may also have to do with the fact that it seems that finetuning is a lot less intense training than LoRa and I am not sure how to circumvent that without model collapse or a larger dataset.
